File: ios/Adapters/RekognitionLivenessAdapter.swift
import Foundation
import UIKit
import SwiftUI

#if canImport(FaceLiveness)
import FaceLiveness
#endif

/// Wraps Amplify Swift's `FaceLivenessDetectorView`.
///
/// The Amplify Face Liveness pod (`AmplifyUIFaceLiveness`) is added
/// by the consumer's config plugin entry when they list
/// `"rekognition"` under `providers`. When the pod is absent this
/// adapter compiles as a stub that throws `LIVENESS_NOT_SUPPORTED`,
/// keeping the umbrella module buildable for consumers who do not
/// use Rekognition.
final class RekognitionLivenessAdapter: LivenessAdapter {
  static let providerName = "rekognition"

  func run(bootstrap: [String: Any], timeoutMs: Int?, locale: String?) async throws -> String {
    guard let sessionId = bootstrap["sessionId"] as? String,
          let region = bootstrap["region"] as? String else {
      throw livenessError(
        .invalidBootstrap,
        "Rekognition bootstrap is missing sessionId or region"
      )
    }

    #if canImport(FaceLiveness)
    return try await presentFaceLiveness(sessionId: sessionId, region: region)
    #else
    throw livenessError(
      .notSupported,
      "AmplifyUIFaceLiveness pod is not installed. Add \"rekognition\" to the expo-passkey-liveness config plugin providers list and re-run prebuild."
    )
    #endif
  }

  #if canImport(FaceLiveness)
  @MainActor
  private func presentFaceLiveness(sessionId: String, region: String) async throws -> String {
    return try await withCheckedThrowingContinuation { continuation in
      guard let root = topMostViewController() else {
        continuation.resume(
          throwing: livenessError(.providerError, "No root view controller available")
        )
        return
      }

      // Box so the host VC can be dismissed from inside the closure.
      let hostRef = HostRef()

      let view = FaceLivenessDetectorView(
        sessionID: sessionId,
        region: region,
        isPresented: Binding<Bool>(
          get: { hostRef.host != nil },
          set: { newValue in
            if !newValue {
              hostRef.host?.dismiss(animated: true) { hostRef.host = nil }
            }
          }
        ),
        onCompletion: { result in
          DispatchQueue.main.async {
            hostRef.host?.dismiss(animated: true) { hostRef.host = nil }
            switch result {
            case .success:
              let payload: [String: Any] = [
                "sessionId": sessionId
              ]
              if let data = try? JSONSerialization.data(withJSONObject: payload),
                 let str = String(data: data, encoding: .utf8) {
                continuation.resume(returning: str)
              } else {
                continuation.resume(
                  throwing: livenessError(.providerError, "Failed to serialise completion payload")
                )
              }
            case .failure(let error):
              continuation.resume(
                throwing: Self.mapFaceLivenessError(error)
              )
            }
          }
        }
      )

      let host = UIHostingController(rootView: view)
      host.modalPresentationStyle = .fullScreen
      hostRef.host = host
      root.present(host, animated: true, completion: nil)
    }
  }

  private static func mapFaceLivenessError(_ error: FaceLivenessDetectionError) -> LivenessException {
    switch error {
    case .userCancelled:
      return livenessError(.userCanceled, "User cancelled the liveness check")
    case .accessDenied, .cameraPermissionDenied:
      return livenessError(.cameraPermissionDenied, "Camera permission denied")
    default:
      return livenessError(.providerError, String(describing: error))
    }
  }
  #endif

  private func topMostViewController() -> UIViewController? {
    let scenes = UIApplication.shared.connectedScenes
    let windowScene = scenes.first(where: { $0.activationState == .foregroundActive }) as? UIWindowScene
    let keyWindow = windowScene?.windows.first(where: { $0.isKeyWindow })
    var vc = keyWindow?.rootViewController
    while let presented = vc?.presentedViewController {
      vc = presented
    }
    return vc
  }
}

/// Reference box so we can flip the SwiftUI binding from outside.
private final class HostRef {
  var host: UIViewController?
}

File: android/src/main/java/expo/modules/passkeyliveness/adapters/RekognitionLivenessAdapter.kt
package expo.modules.passkeyliveness.adapters

import android.app.Activity
import android.content.Intent
import expo.modules.passkeyliveness.LivenessErrorCodes
import expo.modules.passkeyliveness.LivenessException
import kotlinx.coroutines.suspendCancellableCoroutine
import org.json.JSONObject
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Wraps Amplify Android's Face Liveness Compose component.
 *
 * The Amplify Liveness Gradle dependency (`com.amplifyframework.ui:liveness`)
 * is added by the consumer's config plugin entry when they list
 * "rekognition" under providers. We use reflection so this file
 * compiles whether or not the SDK is on the classpath; the runtime
 * Class.forName probe surfaces a stable LIVENESS_NOT_SUPPORTED when
 * the consumer has not opted in.
 *
 * The actual Compose composable is hosted by RekognitionLivenessActivity
 * (a small Activity we ship in src/main/java); the consumer must
 * register it in their AndroidManifest.xml via the config plugin.
 */
class RekognitionLivenessAdapter : LivenessAdapter {
  override val providerName: String = PROVIDER_NAME

  override suspend fun run(
    activity: Activity,
    bootstrap: JSONObject,
    timeoutMs: Int?,
    locale: String?,
  ): String {
    if (!isAmplifyLivenessOnClasspath()) {
      throw LivenessException(
        LivenessErrorCodes.NOT_SUPPORTED,
        "com.amplifyframework.ui:liveness is not on the classpath. Add \"rekognition\" to the expo-passkey-liveness config plugin providers list and rebuild."
      )
    }

    val sessionId = bootstrap.optString("sessionId").takeIf { it.isNotEmpty() }
      ?: throw LivenessException(LivenessErrorCodes.INVALID_BOOTSTRAP, "Rekognition bootstrap missing sessionId")
    val region = bootstrap.optString("region").takeIf { it.isNotEmpty() }
      ?: throw LivenessException(LivenessErrorCodes.INVALID_BOOTSTRAP, "Rekognition bootstrap missing region")

    return suspendCancellableCoroutine { cont ->
      val intent = Intent(activity, RekognitionLivenessActivity::class.java).apply {
        putExtra(RekognitionLivenessActivity.EXTRA_SESSION_ID, sessionId)
        putExtra(RekognitionLivenessActivity.EXTRA_REGION, region)
      }
      RekognitionLivenessActivity.pendingContinuation = { result ->
        when (result) {
          is RekognitionLivenessActivity.Result.Success -> {
            val payload = JSONObject().put("sessionId", result.sessionId).toString()
            cont.resume(payload)
          }
          is RekognitionLivenessActivity.Result.Cancelled ->
            cont.resumeWithException(
              LivenessException(LivenessErrorCodes.USER_CANCELED, "User cancelled the liveness check")
            )
          is RekognitionLivenessActivity.Result.PermissionDenied ->
            cont.resumeWithException(
              LivenessException(LivenessErrorCodes.CAMERA_PERMISSION_DENIED, "Camera permission denied")
            )
          is RekognitionLivenessActivity.Result.Error ->
            cont.resumeWithException(
              LivenessException(LivenessErrorCodes.PROVIDER_ERROR, result.message)
            )
        }
      }
      activity.startActivity(intent)
    }
  }

  private fun isAmplifyLivenessOnClasspath(): Boolean = try {
    Class.forName("com.amplifyframework.ui.liveness.ui.FaceLivenessDetector")
    true
  } catch (_: Throwable) {
    false
  }

  companion object {
    const val PROVIDER_NAME = "rekognition"
  }
}
